File: 21.merge-two-sorted-lists.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    pub fn merge_two_lists(
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match (list1, list2) {
            (Some(list1), None) => Some(list1),
            (None, Some(list2)) => Some(list2),
            (None, None) => None,
            (Some(l1), Some(l2)) => {
                if l1.val < l2.val {
                    return Some(Box::new(ListNode {
                        val: l1.val,
                        next: Solution::merge_two_lists(l1.next, Some(l2)),
                    }));
                } else {
                    return Some(Box::new(ListNode {
                        val: l2.val,
                        next: Solution::merge_two_lists(Some(l1), l2.next),
                    }));
                }
            }
        }
    }
}

fn vec_to_list(vec: Vec<i32>) -> Option<Box<ListNode>> {
    let mut current = None;
    for &val in vec.iter().rev() {
        let mut new_node = Box::new(ListNode::new(val));
        new_node.next = current;
        current = Some(new_node);
    }
    current
}

fn list_to_vec(list: Option<Box<ListNode>>) -> Vec<i32> {
    let mut vec = Vec::new();
    let mut current = list;
    while let Some(node) = current {
        vec.push(node.val);
        current = node.next;
    }
    vec
}

// Test functions moved into the global scope
fn test_merge_two_lists_case_1() {
    let list1 = vec_to_list(vec![1, 2, 4]);
    let list2 = vec_to_list(vec![1, 3, 4]);
    let expected = vec![1, 1, 2, 3, 4, 4];
    let result = Solution::merge_two_lists(list1, list2);
    assert_eq!(list_to_vec(result), expected);
    println!("Case 1 succeed 👍")
}

fn test_merge_two_lists_case_2() {
    let list1 = vec_to_list(vec![]);
    let list2 = vec_to_list(vec![]);
    let expected: Vec<i32> = vec![];
    let result = Solution::merge_two_lists(list1, list2);
    assert_eq!(list_to_vec(result), expected);
    println!("Case 2 succeed 👍")
}

fn test_merge_two_lists_case_3() {
    let list1 = vec_to_list(vec![]);
    let list2 = vec_to_list(vec![0]);
    let expected = vec![0];
    let result = Solution::merge_two_lists(list1, list2);
    assert_eq!(list_to_vec(result), expected);
    println!("Case 3 succeed 👍")
}

fn test_merge_two_lists_case_4() {
    let list1 = vec_to_list(vec![5, 10, 15]);
    let list2 = vec_to_list(vec![2, 3, 20]);
    let expected = vec![2, 3, 5, 10, 15, 20];
    let result = Solution::merge_two_lists(list1, list2);
    assert_eq!(list_to_vec(result), expected);
    println!("Case 4 succeed 👍")
}

fn test_merge_two_lists_case_5() {
    let list1 = vec_to_list(vec![2, 3, 7]);
    let list2 = vec_to_list(vec![1, 4, 5]);
    let expected = vec![1, 2, 3, 4, 5, 7];
    let result = Solution::merge_two_lists(list1, list2);
    assert_eq!(list_to_vec(result), expected);
    println!("Case 5 succeed 👍")
}

fn main() {
    // Simulate running tests
    test_merge_two_lists_case_1();
    test_merge_two_lists_case_2();
    test_merge_two_lists_case_3();
    test_merge_two_lists_case_4();
    test_merge_two_lists_case_5();
}

File: 21.merge-two-sorted-lists.py
from typing import Optional
# @leet start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        node = res = ListNode()
        

        while list1 and list2:
            if list1.val < list2.val:
                node.next = list1
                list1 = list1.next
            else:
                node.next = list2
                list2 = list2.next
            node = node.next

        node.next = list1 or list2

        return res.next

        
# @leet end



# Helper functions to convert between list and ListNode
def list_to_nodes(lst):
    if not lst:
        return None
    dummy = ListNode()
    current = dummy
    for value in lst:
        current.next = ListNode(value)
        current = current.next
    return dummy.next

def nodes_to_list(node):
    lst = []
    while node:
        lst.append(node.val)
        node = node.next
    return lst

# Test functions
def test_merge_two_lists_case_1():
    solution = Solution()
    list1 = list_to_nodes([1, 2, 4])
    list2 = list_to_nodes([1, 3, 4])
    expected = [1, 1, 2, 3, 4, 4]
    result = nodes_to_list(solution.mergeTwoLists(list1, list2))
    assert result == expected, f"Expected {expected}, but got {result}"
    print("Case 1 succeed 👍")

def test_merge_two_lists_case_2():
    solution = Solution()
    list1 = list_to_nodes([])
    list2 = list_to_nodes([])
    expected = []
    result = nodes_to_list(solution.mergeTwoLists(list1, list2))
    assert result == expected, f"Expected {expected}, but got {result}"
    print("Case 2 succeed 👍")

def test_merge_two_lists_case_3():
    solution = Solution()
    list1 = list_to_nodes([])
    list2 = list_to_nodes([0])
    expected = [0]
    result = nodes_to_list(solution.mergeTwoLists(list1, list2))
    assert result == expected, f"Expected {expected}, but got {result}"
    print("Case 3 succeed 👍")

def main():
    # Simulate running tests
    test_merge_two_lists_case_1()
    test_merge_two_lists_case_2()
    test_merge_two_lists_case_3()

if __name__ == "__main__":
    main()
